framework/src/state_machine/state_machine.ts

```diff
@@ -194,6 +194,8 @@ export class StateMachine {
 				try {
 					await mod.afterCommandExecute(transactionContext);
 				} catch (error) {
+					ctx.eventQueue.restoreSnapshot(snapshotID);
+					ctx.stateStore.restoreSnapshot();
```

Comment:
if it fails on the `afterCommandExecute` it should revert to before calling `executeTransaction`

Reply:
I think `Snapshot` needs to be created and revert if before/afterCommandExecute fails to ensure Invalid transactions not to change the state.

we could do it in this function or the function using this function

Reply:
Thanks for adding the beforeCommandEecute, I was focussing on the afterCommandExecute only. I have pushed the createSnapshot to before.

Reply:
Ah sorry i think the spec was not so clear. What I meant was
```
const rootSnapshot = xxx.createSnapshot();
try {
beforeCommandExecute()
} catch {
 //  back to root
}
const commandSnapshot = xxx.createSnapshot();
try {
  command.execute()
} catch {
 //  back to command snapshot
}
try {
afterCommandExecute()
} catch {
 //  back to root
}
```